src/configparser/field_coersion.rs

```diff
@@ -0,0 +1,55 @@
+// stuff to coerce bare string into full build context object
+// (based on serde example: https://serde.rs/string-or-struct.html)
+
+use std::collections::BTreeMap as Map;
+use std::fmt;
+use std::marker::PhantomData;
+use std::str::FromStr;
+
+use serde::de::{self, MapAccess, Visitor};
+use serde::{Deserialize, Deserializer};
+use void::Void;
+
+pub fn string_or_struct<'de, T, D>(deserializer: D) -> Result<T, D::Error>
+where
+    T: Deserialize<'de> + FromStr<Err = Void>,
+    D: Deserializer<'de>,
+{
+    // This is a Visitor that forwards string types to T's `FromStr` impl and
+    // forwards map types to T's `Deserialize` impl. The `PhantomData` is to
+    // keep the compiler from complaining about T being an unused generic type
+    // parameter. We need T in order to know the Value type for the Visitor
+    // impl.
+    struct StringOrStruct<T>(PhantomData<fn() -> T>);
+
+    impl<'de, T> Visitor<'de> for StringOrStruct<T>
+    where
+        T: Deserialize<'de> + FromStr<Err = Void>,
+    {
+        type Value = T;
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            formatter.write_str("string or map")
+        }
+
+        fn visit_str<E>(self, value: &str) -> Result<T, E>
+        where
+            E: de::Error,
+        {
+            Ok(FromStr::from_str(value).unwrap())
```

Comment:
unwrap here probably isn't ideal, I'd prefer returning a `de::Error` variant (maybe with [`de::Error::invalid_value`](https://docs.rs/serde/latest/serde/de/trait.Error.html#method.invalid_value)?)

Reply:
`value` here is always going to be a `&str`, so AFAICT there's no chance it will ever fail here. I stole this code from Serde's documentation examples so I'm also not really sure why this is doing the `from_str` dance on an existing `&str` to begin with.